File: dynamodb-sdk/table.js
const AWS = require('aws-sdk')
const uuidv4 = require('uuid/v4')

class Table {
    constructor(table_name) {
        this._table_name = table_name
        this._dynamodb_client = new AWS.DynamoDB.DocumentClient({ region: "us-east-1" })
    }

    getRecords(params) {
        params["TableName"] = this._table_name
        return this._dynamodb_client.scan(params).promise()
    }

    getRecordById(id) {
        const params = this._getParamsForGetRecordById(id)
        return this._dynamodb_client.get(params).promise()
    }

    _getParamsForGetRecordById(id, attributes_to_get) {
        const params = this._getSubParamsWithIdAndTableName(id)
        
        if (attributes_to_get) {
            params.AttributesToGet = attributes_to_get
        }

        return params
    }

    insertRecord(data) {
        const included_id_data = this._injectIdForRecord(data)
        const params = this._getParamsForInsertRecord(included_id_data)
        
        return this._dynamodb_client.put(params).promise().then(() => {
            return included_id_data
        })
    }

    _getParamsForInsertRecord(included_id_data) {
        const params = {
            TableName: this._table_name,
            Item: included_id_data,
            ReturnValues: "ALL_OLD"
        }

        return params
    }

    _injectIdForRecord(record) {
        const id = this._randomId()
        const included_id_record = Object.assign({id: id}, record)

        return included_id_record
    }

    _randomId() {
        return uuidv4()
    }

    deleteRecordById(id) {
        const params = this._getParamsForDeleteRecordById(id)
        return this._dynamodb_client.delete(params).promise()
    }

    _getParamsForDeleteRecordById(id) {
        const params = this._getSubParamsWithIdAndTableName(id)
        return params
    }

    updateRecordById(id, new_data) {
        const params = this._getParamsForUpdateRecordById(id, new_data)
        return this._dynamodb_client.update(params).promise()
    }

    _getParamsForUpdateRecordById(id, new_data) {
        const params = this._getSubParamsWithIdAndTableName(id)
        const update_expression = this._getUpdateExpressionForUpdateRecordById(new_data)
        const expression_attribute_names = this._getExpressionAttributeNamesForUpdateRecordById(new_data)
        const expression_attribute_values = this._getExpressionAttributeValuesForUpdateRecordById(new_data)

        params.UpdateExpression = update_expression
        params.ExpressionAttributeNames = expression_attribute_names
        params.ExpressionAttributeValues = expression_attribute_values
        params.ReturnValues = "ALL_NEW"

        return params
    }

    _getUpdateExpressionForUpdateRecordById(new_data) {
        return "set " + Object.keys(new_data).map(key => `#${key}=:${key}`).join(",")
    }

    _getExpressionAttributeNamesForUpdateRecordById(new_data) {
        const result = {}
        Object.keys(new_data).forEach(key => {
            result[`#${key}`] = key
        })

        return result
    }

    _getExpressionAttributeValuesForUpdateRecordById(new_data) {
        const result = {}
        Object.keys(new_data).forEach(key => {
            result[`:${key}`] = new_data[key]
        })

        return result
    }

    deleteRecordsByIds(ids) {
        const params = this._getParamsForDeleteRecordsByIds(ids)
        return this._dynamodb_client.batchWrite(params).promise()
    }

    _getParamsForDeleteRecordsByIds(ids) {
        const request_items = {}
        request_items[this._table_name] = ids.map(id => {
            return {
                DeleteRequest: {
                    Key: {
                        id: id
                    }
                }
            }
        })
        const params = {
            RequestItems: request_items,
            ReturnConsumedCapacity: "TOTAL"
        }

        return params
    }

    insertRecords(records) {
        const params = this._getParamForInsertRecords(records)
        return this._dynamodb_client.batchWrite(params).promise()
    }

    _getParamForInsertRecords(records) {
        const included_id_records = records.map(record => this._injectIdForRecord(record))
        const request_items = {}
        request_items[this._table_name] = included_id_records.map(included_id_record => {
            return {
                PutRequest: {
                    Item: included_id_record
                }
            }
        })

        const params = {
            RequestItems: request_items,
            ReturnConsumedCapacity: "TOTAL"
        }

        return params
    }

    _getSubParamsWithIdAndTableName(id) {
        const params = {
            TableName: this._table_name,
            Key: {
                id: id
            }
        }

        return params
    }
}

module.exports = Table

File: dynamodb-sdk-python/table.py
import boto3
import json
from uuid import uuid4
from dynamodb_json import json_util

class Table:
    def __init__(self, table_name):
        self._table_name = table_name
        self._dynamodb = boto3.client("dynamodb", region_name="us-east-1")

    def getRecords(self, params):
        params["TableName"] = self._table_name
        result = self._dynamodb.scan(**params)
        return json_util.loads(result["Items"])

    def getRecordById(self, id):
        params = self._getParamsForGetRecordById(id)
        result = self._dynamodb.get_item(**params)
        return json_util.loads(result["Item"])

    def _getParamsForGetRecordById(self, id, attributes_to_get=None):
        params = self._getSubParamsWithIdAndTableName(id)
        
        if (attributes_to_get):
            params["AttributesToGet"] = attributes_to_get

        return params
    

    def _getSubParamsWithIdAndTableName(self, id):
        params = {
            "TableName": self._table_name,
            "Key": {
                "id": {
                    "S": id
                }
            }
        }

        return params
    
    def _randomId(self):
        return str(uuid4())
    
    def insertRecord(self, data):
        included_id_data = self._injectIdForRecord(data)
        params = self._getParamsForInsertRecord(included_id_data)
        
        self._dynamodb.put_item(**params)
        return included_id_data
    
    def _getParamsForInsertRecord(self, included_id_data):
        params = {
            "TableName": self._table_name,
            "Item": json.loads(json_util.dumps(included_id_data)),
            "ReturnValues": "ALL_OLD"
        }

        return params
    
    def _injectIdForRecord(self, record):
        id = self._randomId()
        included_id_record = {
            **record,
            "id": id
        }

        return included_id_record
    
    def deleteRecordById(self, id):
        params = self._getParamsForDeleteRecordById(id)
        
        try:
            self._dynamodb.delete_item(**params)
            return True
        except:
            return False
    

    def _getParamsForDeleteRecordById(self, id):
        params = self._getSubParamsWithIdAndTableName(id)
        return params
    
    def deleteRecordsByIds(self, ids):
        params = self._getParamsForDeleteRecordsByIds(ids)
        
        try:
            self._dynamodb.batch_write_item(**params)
            return True
        except:
            return False
    

    def _getParamsForDeleteRecordsByIds(self, ids):
        request_items = {}

        def _getParamsForDeleteRecord(id):
            return {
                "DeleteRequest": {
                    "Key": {
                        "id": {
                            "S": id
                        }
                    }
                }
            }

        request_items[self._table_name] = list(map(_getParamsForDeleteRecord, ids))
        params = {
            "RequestItems": request_items,
            "ReturnConsumedCapacity": "TOTAL"
        }

        return params
    
    def insertRecords(self, records):
        included_id_records = list(map(lambda record: self._injectIdForRecord(record), records))
        params = self._getParamForInsertRecords(included_id_records)
        self._dynamodb.batch_write_item(**params)

        return list(map(lambda record: record["id"], included_id_records))

    def _getParamForInsertRecords(self, included_id_records):
        request_items = {}

        def _getParamsForInsertRecord(included_id_record):
            return {
                "PutRequest": {
                    "Item": json.loads(json_util.dumps(included_id_record))
                }
            }

        request_items[self._table_name] = list(map(_getParamsForInsertRecord, included_id_records))
        params = {
            "RequestItems": request_items,
            "ReturnConsumedCapacity": "TOTAL"
        }

        return params
    
    def updateRecordById(self, id, new_data):
        params = self._getParamsForUpdateRecordById(id, new_data)
        result = self._dynamodb.update_item(**params)
        return json_util.loads(result["Attributes"])

    def _getParamsForUpdateRecordById(self, id, new_data):
        params = self._getSubParamsWithIdAndTableName(id)
        update_expression = self._getUpdateExpressionForUpdateRecordById(new_data)
        expression_attribute_names = self._getExpressionAttributeNamesForUpdateRecordById(new_data)
        expression_attribute_values = self._getExpressionAttributeValuesForUpdateRecordById(new_data)

        params["UpdateExpression"] = update_expression
        params["ExpressionAttributeNames"] = expression_attribute_names
        params["ExpressionAttributeValues"] = expression_attribute_values
        params["ReturnValues"] = "ALL_NEW"

        return params
    

    def _getUpdateExpressionForUpdateRecordById(self, new_data):
        return "SET " + ",".join(list(map(lambda key: "#{key}=:{key}".format(key=key), new_data.keys())))
    

    def _getExpressionAttributeNamesForUpdateRecordById(self, new_data):
        result = {}
        for key in new_data.keys():
            result["#{key}".format(key=key)] = key
    
        return result
    

    def _getExpressionAttributeValuesForUpdateRecordById(self, new_data):
        result = {}
        for key in new_data.keys():
            result[":{key}".format(key=key)] = new_data[key]
        return json.loads(json_util.dumps(result))
    
    

    